shark_turbine/kernel/wave/utils.py

```diff
@@ -181,3 +190,118 @@ def remove_global_indexing(
         for constraint in tilingConstraints:
             new_index[key] = new_index[key].subs({constraint.induction_var: 0})
     return new_index
+
+
+def _invoke(vm_context, device, entry_function, inputs, outputs):
+    arg_list = rt.VmVariantList(len(inputs))
+    ret_list = rt.VmVariantList(len(outputs))
+
+    for input in inputs:
+        input_cpu = input.cpu().contiguous()
+        device_array = rt.asdevicearray(device, input_cpu)
+        arg_list.push_ref(device_array._buffer_view)
+
+    vm_context.invoke(entry_function, arg_list, ret_list)
+
+    for i, ret in enumerate(outputs):
+        device_buffer_view = rt.HalBufferView.__iree_vm_cast__(ret_list.get_as_ref(i))
+        device_array = rt.DeviceArray(device, device_buffer_view)
+
+        # TODO: Make to_host accept out array/buffer, so we can avoid extra data copy.
+        host_array = device_array.to_host()
+
+        # Convert to torch tensor without actually importing torch.
+        ret[:] = type(ret)(host_array)
+
+
+def _write_file(name, mode, data):
+    with open(name, mode) as file:
+        file.write(data)
+
+
+def _print_bench_result(result, filename):
+    import json
+
+    res = json.dumps(result, sort_keys=True, indent=4)
+    if filename is not None:
+        _write_file(filename, "w", res)
+    else:
+        print(res)
+
+
+def compile_and_invoke(
+    asm: str,
+    func_name: str,
+    config: dict[str, str],
+    kernel_inputs: list[torch.Tensor],
+    kernel_outputs: list[torch.Tensor],
+    run: bool = False,
+    run_bench: bool = False,
+):
+    backend = config["backend"]
+    device = config["device"]
+    flags = [
```

Comment:
Just curious do we not need to specify a different flag when we start compilation from distributed/lower level  vector dialect VS linalg? :O

Reply:
Surprisingly not :) You can call it good design on the iree side :)

Reply:
Nice! :D